File: output/deepseek_r1/avatar/direct_translation/python/java/atcoder_ABC150_E.java

import java.util.*;

public class atcoder_ABC150_E {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        scanner.nextLine();
        String[] cStr = scanner.nextLine().split(" ");
        int[] c = new int[cStr.length];
        for (int i = 0; i < cStr.length; i++) {
            c[i] = Integer.parseInt(cStr[i]);
        }
        Arrays.sort(c);
        reverse(c);
        long mod = 1000000007;
        long exponent = 2L * n - 2;
        long b = powMod(2, exponent, mod);
        long a = (2 * b) % mod;
        long ans = 0;
        for (int i = 0; i < n; i++) {
            long part = (a + i * b) % mod;
            part = (part * c[i]) % mod;
            ans = (ans + part) % mod;
        }
        System.out.println(ans);
    }

    private static void reverse(int[] array) {
        for (int i = 0; i < array.length / 2; i++) {
            int temp = array[i];
            array[i] = array[array.length - 1 - i];
            array[array.length - 1 - i] = temp;
        }
    }

File: output/deepseek_r1/avatar/direct_translation/java/python/atcoder_ABC150_E.py

def power(x, n, mod):
    if n == 0:
        return 1
    val = power(x, n // 2, mod)
    val = val * val % mod
    if n % 2 == 1:
        val = val * x % mod
    return val

n = int(input())
c = list(map(int, input().split()))
mod = 10**9 + 7

if n == 1:
    print((c[0] * 2) % mod)
    exit()

c.sort()
c = c[::-1]

b = power(2, n, mod)
a = power(2, n - 2, mod)

ans = 0
for i in range(2, n + 2):
    val = a * i % mod
    val = val * c[i - 2] % mod
    ans = (ans + val) % mod

ans = ans * b % mod
print(ans)

